File: frontends/macos/keypad/KeypadWindow.m
/*
 * A2600KeypadWindow -- the AppKit keypad panel: both keyboard controllers
 * side by side, each a 3x4 keypad centred over its port's fire buttons,
 * then the console switches and the Map row.
 *
 * Buttons are a PadButton subclass that reports mouseDown and mouseUp
 * rather than an action on click: a keypad key on this machine is HELD, and
 * a game polls it, so a value present only for the instant of a click
 * falls between frames. mouseUp arrives even when the pointer has left the
 * button (AppKit tracks the drag for the view that got mouseDown), so
 * dragging off cannot strand the machine with a key held forever.
 *
 * A utility panel of fixed size (no resize mask): the keys are a block of
 * fixed-size buttons, and there is nothing to gain by stretching them.
 * Singleton, ordered out rather than closed, so its position survives.
 *
 * Copyright (C) 2026 Thomas Cherryhomes
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "KeypadWindow.h"

#import "../KeyForward.h"

/* -2 idle, -1 armed and waiting for a target, >= 0 waiting for a key or
 * pad button. */
static int g_mapState = -2;
static A2600KeypadWindow *g_singleton;
static a2600session *g_session;

/* NOT named `target`: NSControl already has a `target` property. */
@interface PadButton : NSButton
@property (nonatomic) int padTarget;
@property (nonatomic, copy) NSString *face;
@property (nonatomic) BOOL down;
@end

@implementation PadButton
- (void)setHeld:(BOOL)held
{
    self.bezelColor = held ? A2600AccentColor() : nil;
    [self highlight:held];
}

- (void)mouseDown:(NSEvent *)e
{
    (void)e;
    if (g_mapState == -1) {
        g_mapState = self.padTarget;
        [[NSNotificationCenter defaultCenter] postNotificationName:@"A2600PadRefresh" object:nil];
        return;
    }
    if (g_mapState >= 0) return;
    self.down = YES;
    [self setHeld:YES];
    if (self.padTarget >= A2600_TARGET_SYSACT(0)) return;   /* fires on release */
    a2600session_press(g_session, self.padTarget, 1);
}

- (void)mouseUp:(NSEvent *)e
{
    (void)e;
    if (!self.down) return;
    self.down = NO;
    [self setHeld:NO];
    if (g_mapState != -2) return;
    if (self.padTarget >= A2600_TARGET_SYSACT(0)) {
        /* System actions fire on release, like a real button: pressing and
         * dragging off must not reboot the machine. */
        a2600session_sysaction(g_session, self.padTarget - A2600_TARGET_SYSACT(0));
        return;
    }
    a2600session_press(g_session, self.padTarget, 0);
}
@end

@implementation A2600KeypadWindow {
    NSMutableArray<PadButton *> *_buttons;
    NSButton *_mapButton;
    NSTextField *_hint;
    NSTextField *_typeLabel[2];
    NSTimer *_captureTimer;
    NSTimer *_typeTimer;
}

static NSString *const kFace[12] = { @"1", @"2", @"3", @"4", @"5", @"6", @"7", @"8", @"9", @"*", @"0", @"#" };

#define KEY_W   56.0
#define KEY_H   40.0
#define GAP      6.0
#define WIDE_W 118.0
#define PAD_W  (3 * WIDE_W + 2 * GAP)
#define GRID_W (3 * KEY_W + 2 * GAP)
#define MARGIN  12.0

- (PadButton *)buttonWithFace:(NSString *)face target:(int)target frame:(NSRect)frame
{
    PadButton *b = [[PadButton alloc] initWithFrame:frame];
    [b setTitle:face];
    [b setBezelStyle:NSBezelStyleRounded];
    b.padTarget = target;
    b.face = face;
    /* Not focusable: clicking a pad button must not steal the key window's
     * first responder. */
    [b setRefusesFirstResponder:YES];
    [_buttons addObject:b];
    return b;
}

/* One controller, laid out downwards from topY (AppKit's y grows upwards,
 * so every row is placed by its bottom edge). Returns the y below it. */
- (CGFloat)buildController:(int)port intoView:(NSView *)parent x:(CGFloat)x0 topY:(CGFloat)topY
{
    CGFloat y = topY;

    NSTextField *title = [NSTextField labelWithString:(port ? @"Right Port" : @"Left Port")];
    [title setAlignment:NSTextAlignmentCenter];
    [title setFont:[NSFont boldSystemFontOfSize:12]];
    [title setFrame:NSMakeRect(x0, y - 18, PAD_W, 16)];
    [parent addSubview:title];
    y -= 20;

    _typeLabel[port] = [NSTextField labelWithString:@""];
    [_typeLabel[port] setAlignment:NSTextAlignmentCenter];
    [_typeLabel[port] setFont:[NSFont systemFontOfSize:NSFont.smallSystemFontSize]];
    [_typeLabel[port] setFrame:NSMakeRect(x0, y - 16, PAD_W, 14)];
    [parent addSubview:_typeLabel[port]];
    y -= 16 + GAP;

    /* The 3x4 block, centred over the fire row so the digits sit tight. */
    const CGFloat gx = x0 + (PAD_W - GRID_W) / 2;
    for (int i = 0; i < 12; i++) {
        NSRect r = NSMakeRect(gx + (i % 3) * (KEY_W + GAP), y - (i / 3 + 1) * (KEY_H + GAP) + GAP, KEY_W, KEY_H);
        [parent addSubview:[self buttonWithFace:kFace[i]
                                         target:A2600_TARGET_PORT(port, A2600_ACT_KEY_1 + i)
                                          frame:r]];
    }
    y -= 4 * (KEY_H + GAP);

    [parent addSubview:[self buttonWithFace:@"Fire" target:A2600_TARGET_PORT(port, A2600_ACT_JOY_FIRE)
                                      frame:NSMakeRect(x0, y - KEY_H, WIDE_W, KEY_H)]];
    [parent addSubview:[self buttonWithFace:@"Paddle A" target:A2600_TARGET_PORT(port, A2600_ACT_PADDLE_A_FIRE)
                                      frame:NSMakeRect(x0 + WIDE_W + GAP, y - KEY_H, WIDE_W, KEY_H)]];
    [parent addSubview:[self buttonWithFace:@"Paddle B" target:A2600_TARGET_PORT(port, A2600_ACT_PADDLE_B_FIRE)
                                      frame:NSMakeRect(x0 + 2 * (WIDE_W + GAP), y - KEY_H, WIDE_W, KEY_H)]];
    return y - KEY_H;
}

- (instancetype)init
{
    const CGFloat x1 = MARGIN + PAD_W + 2 * MARGIN;
    const CGFloat width = x1 + PAD_W + MARGIN;
    const CGFloat height = MARGIN + 20 + 16 + GAP + 4 * (KEY_H + GAP) + KEY_H   /* a controller */
                         + MARGIN + 4 + KEY_H + MARGIN + 30 + MARGIN;           /* console + map rows */

    NSPanel *win = [[NSPanel alloc]
        initWithContentRect:NSMakeRect(0, 0, width, height)
                  styleMask:(NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskUtilityWindow)
                    backing:NSBackingStoreBuffered
                      defer:NO];
    [win setTitle:@"Keypads"];
    /* Floats above the machine's window and stays out of the way. */
    [win setLevel:NSFloatingWindowLevel];
    [win setReleasedWhenClosed:NO];
    [win setBecomesKeyOnlyIfNeeded:NO];

    self = [super initWithWindow:win];
    if (!self) return nil;
    [win setDelegate:self];

    _buttons = [NSMutableArray array];
    NSView *content = [win contentView];
    CGFloat y = height - MARGIN;
    [self buildController:0 intoView:content x:MARGIN topY:y];
    CGFloat bottom = [self buildController:1 intoView:content x:x1 topY:y];

    /* Console row, centred. */
    bottom -= MARGIN + 4;
    static const struct { NSString *__unsafe_unretained face; int target; } console[6] = {
        { @"Select",           A2600_TARGET_SWITCH(A2600_SW_SELECT) },
        { @"Reset",            A2600_TARGET_SWITCH(A2600_SW_RESET) },
        { @"Color / B&W",      A2600_TARGET_SWITCH(A2600_SW_COLOR_BW) },
        { @"Left Diff",        A2600_TARGET_SWITCH(A2600_SW_LEFT_DIFF) },
        { @"Right Diff",       A2600_TARGET_SWITCH(A2600_SW_RIGHT_DIFF) },
        { @"Reboot to CONFIG", A2600_TARGET_SYSACT(A2600_SYSACT_REBOOT_CONFIG) },
    };
    const CGFloat cx = (width - (6 * WIDE_W + 5 * GAP)) / 2;
    for (int i = 0; i < 6; i++)
        [content addSubview:[self buttonWithFace:console[i].face target:console[i].target
                                           frame:NSMakeRect(cx + i * (WIDE_W + GAP), bottom - KEY_H, WIDE_W, KEY_H)]];
    bottom -= KEY_H + MARGIN;

    _mapButton = [NSButton buttonWithTitle:@"Map" target:self action:@selector(toggleMap:)];
    [_mapButton setFrame:NSMakeRect(MARGIN, bottom - 30, 70, 30)];
    [_mapButton setRefusesFirstResponder:YES];
    [content addSubview:_mapButton];

    NSButton *defaults = [NSButton buttonWithTitle:@"Defaults" target:self action:@selector(restoreDefaults:)];
    [defaults setFrame:NSMakeRect(MARGIN + 76, bottom - 30, 90, 30)];
    [defaults setRefusesFirstResponder:YES];
    [content addSubview:defaults];

    _hint = [NSTextField labelWithString:@""];
    [_hint setFrame:NSMakeRect(MARGIN + 176, bottom - 24, width - MARGIN - 190, 18)];
    [_hint setTextColor:[NSColor secondaryLabelColor]];
    [content addSubview:_hint];

    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refresh)
                                                 name:@"A2600PadRefresh" object:nil];
    [self refresh];
    return self;
}

- (void)toggleMap:(id)sender
{
    (void)sender;
    [self setMapState:(g_mapState == -2) ? -1 : -2];
}

- (void)restoreDefaults:(id)sender
{
    (void)sender;
    a2600session_bindings_reset(g_session);
    [self refresh];
}

- (void)setMapState:(int)state
{
    g_mapState = state;
    [_captureTimer invalidate];
    _captureTimer = nil;
    if (state >= 0) {
        a2600session_gamepad_capture_begin(g_session);
        _captureTimer = [NSTimer scheduledTimerWithTimeInterval:0.05 repeats:YES block:^(NSTimer *t) {
            (void)t;
            [self pollCapture];
        }];
    } else {
        a2600session_gamepad_capture_cancel(g_session);
    }
    [self refresh];
}

- (void)pollCapture
{
    int button;
    if (g_mapState < 0) return;
    if (a2600session_gamepad_capture_poll(g_session, &button)) {
        char stolen[128];
        a2600session_binding_set_button(g_session, g_mapState, button, stolen, sizeof stolen);
        [self setMapState:-1];
        if (stolen[0])
            _hint.stringValue = [NSString stringWithFormat:@"Bound %s (was %s)", a2600_pad_button_name(button), stolen];
    }
}

- (void)refresh
{
    [_mapButton setTitle:(g_mapState == -2 ? @"Map" : @"Cancel")];
    _mapButton.bezelColor = g_mapState == -2 ? nil : A2600AccentColor();
    if (g_mapState == -2)
        [_hint setStringValue:@""];
    else if (g_mapState == -1)
        [_hint setStringValue:@"Click a control to remap"];
    else
        [_hint setStringValue:[NSString stringWithFormat:@"Press a key or gamepad button for %s",
                               a2600_target_name(g_mapState)]];

    for (PadButton *b in _buttons) {
        if (g_mapState != -2) {
            const a2600_binding bind = a2600session_binding_get(g_session, b.padTarget);
            char key[32];
            a2600session_keysym_name(bind.keysym, key, sizeof key);
            NSString *text = key[0] ? [NSString stringWithUTF8String:key] : @"—";
            if (bind.button != A2600_PAD_BTN_NONE)
                text = [text stringByAppendingFormat:@" / %s", a2600_pad_button_name(bind.button)];
            [b setTitle:text];
            b.bezelColor = (b.padTarget == g_mapState) ? A2600AccentColor() : nil;
        } else {
            [b setTitle:b.face];
            if (!b.down) b.bezelColor = nil;
        }
    }
}

- (void)refreshTypes
{
    for (int port = 0; port < 2; port++) {
        const int det = a2600session_detected_port_type(g_session, port);
        _typeLabel[port].stringValue = [NSString stringWithFormat:@"%s attached", a2600_ctrl_type_name(det)];
        _typeLabel[port].textColor = det == A2600_CTRL_KEYPAD ? A2600AccentColor() : [NSColor secondaryLabelColor];
    }
}

/* Keyboard here behaves exactly as in the main window, so typing drives the
 * machine whichever window is key. */
- (void)keyDown:(NSEvent *)e
{
    if ([e isARepeat]) return;
    const uint32_t ks = A2600KeysymFromEvent(e);
    if (g_mapState >= 0) {
        if (ks) {
            char stolen[128], name[32];
            a2600session_binding_set_key(g_session, g_mapState, ks, stolen, sizeof stolen);
            a2600session_keysym_name(ks, name, sizeof name);
            [self setMapState:-1];   /* stay armed: remapping several in a row is normal */
            if (stolen[0])
                _hint.stringValue = [NSString stringWithFormat:@"Bound %s (was %s)", name, stolen];
        }
        return;
    }
    if (g_mapState == -1 || !ks) return;
    if (ks == A2600_KEYSYM_F9) { [A2600KeypadWindow toggleWithSession:g_session]; return; }

    const int sa = a2600session_key_sysaction(g_session, ks);
    if (sa >= 0) { a2600session_sysaction(g_session, sa); return; }
    a2600session_key(g_session, ks, 1);
}

- (void)keyUp:(NSEvent *)e
{
    if (g_mapState != -2) return;
    const uint32_t ks = A2600KeysymFromEvent(e);
    if (ks) a2600session_key(g_session, ks, 0);
}

- (void)flagsChanged:(NSEvent *)e
{
    int down = 0;
    const uint32_t ks = A2600KeysymFromFlagsChange(e, &down);
    if (ks && g_mapState == -2) a2600session_key(g_session, ks, down);
}

- (void)windowWillClose:(NSNotification *)note
{
    (void)note;
    [self setMapState:-2];
    [_typeTimer invalidate];
    _typeTimer = nil;
    a2600session_release_all(g_session);
}

+ (void)toggleWithSession:(a2600session *)session
{
    g_session = session;
    if (!g_singleton) g_singleton = [[A2600KeypadWindow alloc] init];

    if ([[g_singleton window] isVisible]) {
        [g_singleton setMapState:-2];
        [g_singleton->_typeTimer invalidate];
        g_singleton->_typeTimer = nil;
        [[g_singleton window] orderOut:nil];
    } else {
        [g_singleton refreshTypes];
        g_singleton->_typeTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 repeats:YES block:^(NSTimer *t) {
            (void)t;
            [g_singleton refreshTypes];
        }];
        [[g_singleton window] makeKeyAndOrderFront:nil];
    }
}

+ (BOOL)isVisible
{
    return g_singleton && [[g_singleton window] isVisible];
}
@end

File: frontends/windows/keypad/keypad_window.c
/*
 * The Win32 keypad panel: both keyboard controllers side by side, each a
 * 3x4 keypad over its port's fire buttons, then the console switches and
 * the Map row.
 *
 * Buttons are driven by WM_LBUTTONDOWN/WM_LBUTTONUP on the panel window
 * rather than by BN_CLICKED: a keypad key on this machine is HELD, and a
 * game polls it, so a value present only for the instant of a click falls
 * between frames. The mouse is captured on press and released on
 * button-up wherever that happens, so dragging off a button cannot strand
 * the machine with a key held forever.
 *
 * The window is a tool window of fixed size: the keys are a block of
 * fixed-size buttons, and there is nothing to gain by stretching them.
 *
 * Copyright (C) 2026 Thomas Cherryhomes
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "keypad_window.h"

/* GET_X_LPARAM / GET_Y_LPARAM live here, not in windows.h. */
#include <windowsx.h>

#include <stdio.h>
#include <string.h>

#include "../key_forward.h"

#define PAD_CLASS "FujiNetGoAtari2600Keypad"

#define KEY_W   56
#define KEY_H   40
#define GAP      6
#define WIDE_W 118
#define PAD_W  (3 * WIDE_W + 2 * GAP)      /* a controller box's inner width */
#define GRID_W (3 * KEY_W + 2 * GAP)       /* the keypad block's width */
#define MARGIN  12
#define HEAD_H  22                          /* "Left Port" */
#define TYPE_H  20                          /* "Joystick attached" */

#define IDT_CAPTURE 1
#define IDT_TYPES   2

typedef struct {
    RECT rc;
    int target;
    char face[20];
} pad_button;

static HWND g_panel;
static a2600session *g_session;
static pad_button g_btn[A2600_TARGET_COUNT];
static int g_nbtn;
static int g_held = -1;          /* button index under the captured mouse */
static int g_map_state = -2;     /* -2 idle, -1 armed, >=0 awaiting a key/button */
static RECT g_map_rc, g_defaults_rc, g_hint_rc;
static RECT g_head_rc[2], g_type_rc[2];
static int g_detected[2];
static char g_hint[160];
static HBRUSH g_accent_brush;
static HFONT g_bold;

static void add_button(const char *face, int target, int x, int y, int w)
{
    pad_button *b;
    if (g_nbtn >= A2600_TARGET_COUNT) return;
    b = &g_btn[g_nbtn++];
    SetRect(&b->rc, x, y, x + w, y + KEY_H);
    b->target = target;
    snprintf(b->face, sizeof b->face, "%s", face);
}

static const char *const kFace[12] = { "1", "2", "3", "4", "5", "6", "7", "8", "9", "*", "0", "#" };

/* One controller: heading, attached-type line, the centred 3x4 block, and
 * the fire row. Returns the y below it. */
static int build_controller(int port, int x0, int y0)
{
    const int gx = x0 + (PAD_W - GRID_W) / 2;   /* centre the block over the fire row */
    int y = y0, i;

    SetRect(&g_head_rc[port], x0, y, x0 + PAD_W, y + HEAD_H);
    y += HEAD_H;
    SetRect(&g_type_rc[port], x0, y, x0 + PAD_W, y + TYPE_H);
    y += TYPE_H + GAP;

    for (i = 0; i < 12; i++)
        add_button(kFace[i], A2600_TARGET_PORT(port, A2600_ACT_KEY_1 + i),
                   gx + (i % 3) * (KEY_W + GAP), y + (i / 3) * (KEY_H + GAP), KEY_W);
    y += 4 * (KEY_H + GAP) + GAP;

    add_button("Fire", A2600_TARGET_PORT(port, A2600_ACT_JOY_FIRE), x0, y, WIDE_W);
    add_button("Paddle A", A2600_TARGET_PORT(port, A2600_ACT_PADDLE_A_FIRE), x0 + WIDE_W + GAP, y, WIDE_W);
    add_button("Paddle B", A2600_TARGET_PORT(port, A2600_ACT_PADDLE_B_FIRE), x0 + 2 * (WIDE_W + GAP), y, WIDE_W);
    return y + KEY_H;
}

static int g_total_w, g_total_h;

static void layout(void)
{
    const int x1 = MARGIN + PAD_W + 2 * MARGIN;
    int y, cx;
    static const struct { const char *face; int target; } console[6] = {
        { "Select",           A2600_TARGET_SWITCH(A2600_SW_SELECT) },
        { "Reset",            A2600_TARGET_SWITCH(A2600_SW_RESET) },
        { "Color / B&W",      A2600_TARGET_SWITCH(A2600_SW_COLOR_BW) },
        { "Left Diff",        A2600_TARGET_SWITCH(A2600_SW_LEFT_DIFF) },
        { "Right Diff",       A2600_TARGET_SWITCH(A2600_SW_RIGHT_DIFF) },
        { "Reboot to CONFIG", A2600_TARGET_SYSACT(A2600_SYSACT_REBOOT_CONFIG) },
    };
    int i;

    g_nbtn = 0;
    y = build_controller(0, MARGIN, MARGIN);
    build_controller(1, x1, MARGIN);
    g_total_w = x1 + PAD_W + MARGIN;

    /* Console row, centred. */
    y += MARGIN + 4;
    cx = (g_total_w - (6 * WIDE_W + 5 * GAP)) / 2;
    for (i = 0; i < 6; i++)
        add_button(console[i].face, console[i].target, cx + i * (WIDE_W + GAP), y, WIDE_W);
    y += KEY_H + MARGIN;

    SetRect(&g_map_rc, MARGIN, y, MARGIN + 70, y + 30);
    SetRect(&g_defaults_rc, MARGIN + 76, y, MARGIN + 76 + 84, y + 30);
    SetRect(&g_hint_rc, MARGIN + 76 + 84 + 10, y, g_total_w - MARGIN, y + 30);
    g_total_h = y + 30 + MARGIN;
}

static void press_target(int target, int down)
{
    if (target >= A2600_TARGET_SYSACT(0)) {
        /* System actions fire on release, like a real button: pressing and
         * dragging off must not reboot the machine. */
        if (!down) a2600session_sysaction(g_session, target - A2600_TARGET_SYSACT(0));
        return;
    }
    a2600session_press(g_session, target, down);
}

static int hit(int x, int y)
{
    POINT p = { x, y };
    int i;
    for (i = 0; i < g_nbtn; i++)
        if (PtInRect(&g_btn[i].rc, p)) return i;
    return -1;
}

static void set_map_state(int state)
{
    g_map_state = state;
    if (state == -2) {
        KillTimer(g_panel, IDT_CAPTURE);
        a2600session_gamepad_capture_cancel(g_session);
        g_hint[0] = '\0';
    } else if (state == -1) {
        KillTimer(g_panel, IDT_CAPTURE);
        a2600session_gamepad_capture_cancel(g_session);
        snprintf(g_hint, sizeof g_hint, "Click a control to remap");
    } else {
        snprintf(g_hint, sizeof g_hint, "Press a key or gamepad button for %s", a2600_target_name(state));
        a2600session_gamepad_capture_begin(g_session);
        SetTimer(g_panel, IDT_CAPTURE, 50, NULL);
    }
    InvalidateRect(g_panel, NULL, TRUE);
}

static void button_label(const pad_button *b, char *out, int outsz)
{
    if (g_map_state != -2) {
        const a2600_binding bind = a2600session_binding_get(g_session, b->target);
        char key[32];
        a2600session_keysym_name(bind.keysym, key, sizeof key);
        if (bind.button != A2600_PAD_BTN_NONE)
            snprintf(out, outsz, "%s\n%s", key[0] ? key : "-", a2600_pad_button_name(bind.button));
        else
            snprintf(out, outsz, "%s", key[0] ? key : "-");
    } else {
        snprintf(out, outsz, "%s", b->face);
    }
}

static void draw_button(HDC dc, const RECT *rc, const char *label, int pushed, int accent)
{
    RECT r = *rc, text;
    int h;

    if (accent) {
        FillRect(dc, &r, g_accent_brush);
        FrameRect(dc, &r, (HBRUSH)GetStockObject(GRAY_BRUSH));
        SetTextColor(dc, RGB(0, 0, 0));
    } else {
        DrawFrameControl(dc, &r, DFC_BUTTON, DFCS_BUTTONPUSH | (pushed ? DFCS_PUSHED : 0));
        SetTextColor(dc, GetSysColor(COLOR_BTNTEXT));
    }
    /* Vertically centred, word-broken on the newline a Map-mode label
     * carries between its key and its pad button. */
    text = r;
    InflateRect(&text, -3, 0);
    h = DrawTextA(dc, label, -1, &text, DT_CENTER | DT_WORDBREAK | DT_CALCRECT | DT_NOPREFIX);
    text.left = r.left + 3;
    text.right = r.right - 3;
    text.top = r.top + ((r.bottom - r.top) - h) / 2 + (pushed ? 1 : 0);
    text.bottom = r.bottom;
    DrawTextA(dc, label, -1, &text, DT_CENTER | DT_WORDBREAK | DT_NOPREFIX);
}

static void paint_panel(HDC dc)
{
    HFONT font = (HFONT)GetStockObject(DEFAULT_GUI_FONT);
    HGDIOBJ old = SelectObject(dc, font);
    RECT client;
    int i, port;

    GetClientRect(g_panel, &client);
    FillRect(dc, &client, (HBRUSH)(COLOR_BTNFACE + 1));
    SetBkMode(dc, TRANSPARENT);

    for (port = 0; port < 2; port++) {
        char line[64];
        SelectObject(dc, g_bold);
        SetTextColor(dc, GetSysColor(COLOR_BTNTEXT));
        DrawTextA(dc, port ? "Right Port" : "Left Port", -1, &g_head_rc[port], DT_CENTER | DT_VCENTER | DT_SINGLELINE | DT_NOPREFIX);
        SelectObject(dc, g_detected[port] == A2600_CTRL_KEYPAD ? g_bold : font);
        SetTextColor(dc, g_detected[port] == A2600_CTRL_KEYPAD ? RGB(0xC0, 0x60, 0x00) : GetSysColor(COLOR_GRAYTEXT));
        snprintf(line, sizeof line, "%s attached", a2600_ctrl_type_name(g_detected[port]));
        DrawTextA(dc, line, -1, &g_type_rc[port], DT_CENTER | DT_VCENTER | DT_SINGLELINE | DT_NOPREFIX);
    }
    SelectObject(dc, font);

    for (i = 0; i < g_nbtn; i++) {
        char label[64];
        int held = (i == g_held) && g_map_state == -2;
        int accent = held || (g_map_state >= 0 && g_btn[i].target == g_map_state);
        button_label(&g_btn[i], label, sizeof label);
        draw_button(dc, &g_btn[i].rc, label, held, accent);
    }

    draw_button(dc, &g_map_rc, g_map_state == -2 ? "Map" : "Cancel", 0, g_map_state != -2);
    draw_button(dc, &g_defaults_rc, "Defaults", 0, 0);

    if (g_hint[0]) {
        SetTextColor(dc, GetSysColor(COLOR_GRAYTEXT));
        DrawTextA(dc, g_hint, -1, &g_hint_rc, DT_LEFT | DT_VCENTER | DT_SINGLELINE | DT_END_ELLIPSIS | DT_NOPREFIX);
    }
    SelectObject(dc, old);
}

static void refresh_types(void)
{
    int port, changed = 0;
    for (port = 0; port < 2; port++) {
        int det = a2600session_detected_port_type(g_session, port);
        if (det != g_detected[port]) { g_detected[port] = det; changed = 1; }
    }
    if (changed) InvalidateRect(g_panel, NULL, TRUE);
}

static LRESULT CALLBACK pad_proc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp)
{
    switch (msg) {
    case WM_PAINT: {
        PAINTSTRUCT ps;
        HDC dc = BeginPaint(hwnd, &ps);
        paint_panel(dc);
        EndPaint(hwnd, &ps);
        return 0;
    }
    case WM_ERASEBKGND:
        return 1;

    case WM_LBUTTONDOWN: {
        const int x = GET_X_LPARAM(lp), y = GET_Y_LPARAM(lp);
        POINT p = { x, y };
        int i = hit(x, y);

        if (PtInRect(&g_map_rc, p)) { set_map_state(g_map_state == -2 ? -1 : -2); return 0; }
        if (PtInRect(&g_defaults_rc, p)) {
            a2600session_bindings_reset(g_session);
            InvalidateRect(hwnd, NULL, TRUE);
            return 0;
        }
        if (i < 0) return 0;

        if (g_map_state == -1) { set_map_state(g_btn[i].target); return 0; }
        if (g_map_state >= 0) return 0;

        g_held = i;
        SetCapture(hwnd);
        press_target(g_btn[i].target, 1);
        InvalidateRect(hwnd, &g_btn[i].rc, FALSE);
        return 0;
    }
    case WM_LBUTTONUP:
        /* Release wherever the mouse ended up: capture means this arrives
         * even if the pointer left the button. */
        if (g_held >= 0) {
            RECT r = g_btn[g_held].rc;
            int t = g_btn[g_held].target;
            g_held = -1;
            ReleaseCapture();
            press_target(t, 0);
            InvalidateRect(hwnd, &r, FALSE);
        }
        return 0;

    case WM_TIMER:
        if (wp == IDT_CAPTURE && g_map_state >= 0) {
            int button;
            if (a2600session_gamepad_capture_poll(g_session, &button)) {
                char stolen[128];
                a2600session_binding_set_button(g_session, g_map_state, button, stolen, sizeof stolen);
                set_map_state(-1);
                if (stolen[0])
                    snprintf(g_hint, sizeof g_hint, "Bound %s (was %s)", a2600_pad_button_name(button), stolen);
                InvalidateRect(hwnd, NULL, TRUE);
            }
        } else if (wp == IDT_TYPES) {
            refresh_types();
        }
        return 0;

    case WM_KEYDOWN: case WM_SYSKEYDOWN: {
        const uint32_t ks = a2600_keysym_from_msg(wp, lp);
        int sa;
        if (lp & (1 << 30)) return 0;   /* auto-repeat */
        if (g_map_state >= 0) {
            if (ks) {
                char stolen[128], name[32];
                a2600session_binding_set_key(g_session, g_map_state, ks, stolen, sizeof stolen);
                a2600session_keysym_name(ks, name, sizeof name);
                set_map_state(-1);   /* stay armed: remapping several in a row is normal */
                if (stolen[0]) snprintf(g_hint, sizeof g_hint, "Bound %s (was %s)", name, stolen);
                InvalidateRect(hwnd, NULL, TRUE);
            }
            return 0;
        }
        if (g_map_state == -1) return 0;
        if (wp == VK_F9) { ShowWindow(hwnd, SW_HIDE); return 0; }
        if (!ks) break;
        sa = a2600session_key_sysaction(g_session, ks);
        if (sa >= 0) { a2600session_sysaction(g_session, sa); return 0; }
        if (a2600session_key(g_session, ks, 1)) return 0;
        break;
    }
    case WM_KEYUP: case WM_SYSKEYUP: {
        const uint32_t ks = a2600_keysym_from_msg(wp, lp);
        if (g_map_state != -2) return 0;
        if (ks && a2600session_key(g_session, ks, 0)) return 0;
        break;
    }
    case WM_ACTIVATE:
        if (LOWORD(wp) == WA_INACTIVE) a2600session_release_all(g_session);
        return 0;
    case WM_SHOWWINDOW:
        if (wp) { refresh_types(); SetTimer(hwnd, IDT_TYPES, 1000, NULL); }
        else KillTimer(hwnd, IDT_TYPES);
        break;
    case WM_CLOSE:
        /* Hide, do not destroy: the window's position survives closing it. */
        set_map_state(-2);
        ShowWindow(hwnd, SW_HIDE);
        return 0;
    default: break;
    }
    return DefWindowProcA(hwnd, msg, wp, lp);
}

void a2600_keypad_window_toggle(HWND parent, a2600session *session)
{
    g_session = session;

    if (!g_panel) {
        WNDCLASSEXA wc;
        RECT want;
        LOGFONTA lf;

        layout();
        g_accent_brush = CreateSolidBrush(RGB((A2600SESSION_ACCENT_RGB >> 16) & 0xff,
                                              (A2600SESSION_ACCENT_RGB >> 8) & 0xff,
                                              A2600SESSION_ACCENT_RGB & 0xff));
        GetObjectA(GetStockObject(DEFAULT_GUI_FONT), sizeof lf, &lf);
        lf.lfWeight = FW_BOLD;
        g_bold = CreateFontIndirectA(&lf);

        memset(&wc, 0, sizeof wc);
        wc.cbSize = sizeof wc;
        wc.lpfnWndProc = pad_proc;
        wc.hInstance = GetModuleHandle(NULL);
        wc.hCursor = LoadCursor(NULL, IDC_ARROW);
        wc.hbrBackground = (HBRUSH)(COLOR_BTNFACE + 1);
        wc.lpszClassName = PAD_CLASS;
        RegisterClassExA(&wc);

        SetRect(&want, 0, 0, g_total_w, g_total_h);
        /* A tool window with a caption and no thick frame or maximize box:
         * it is exactly the size of its controls. WS_EX_TOOLWINDOW also
         * keeps it off the taskbar. */
        AdjustWindowRectEx(&want, WS_CAPTION | WS_SYSMENU, FALSE, WS_EX_TOOLWINDOW);
        g_panel = CreateWindowExA(WS_EX_TOOLWINDOW, PAD_CLASS, "Keypads",
                                  WS_CAPTION | WS_SYSMENU, CW_USEDEFAULT, CW_USEDEFAULT,
                                  want.right - want.left, want.bottom - want.top,
                                  parent, NULL, wc.hInstance, NULL);
        if (!g_panel) return;
        g_detected[0] = g_detected[1] = -1;
    }

    if (IsWindowVisible(g_panel)) {
        set_map_state(-2);
        ShowWindow(g_panel, SW_HIDE);
    } else {
        ShowWindow(g_panel, SW_SHOW);
        SetForegroundWindow(g_panel);
    }
}

int a2600_keypad_pretranslate(MSG *msg)
{
    /* The panel has no child controls, so its own window proc sees every
     * key; nothing to steal here. Kept for symmetry with the debugger. */
    (void)msg;
    return 0;
}
